src/tqec/sketchup/block_graph.py

```diff
@@ -0,0 +1,551 @@
+"""Graph representation of a 3D spacetime defect diagram by explicit blocks."""
+
+from __future__ import annotations
+
+from io import BytesIO
+import typing as ty
+import pathlib
+from dataclasses import dataclass, astuple
+from enum import Enum
+
+import networkx as nx
+
+from tqec.position import Position3D, Direction3D
+from tqec.sketchup.zx_graph import (
+    ZXGraph,
+    NodeType,
+    ZXNode,
+    ZXEdge,
+)
+from tqec.exceptions import TQECException
+
+
+@dataclass(frozen=True)
+class Color3D:
+    """Get face colors along the x, y, and z axes."""
+
+    x: str | None
+    y: str | None
+    z: str | None
+
+    def __post_init__(self) -> None:
+        if any(c not in "xz" for c in astuple(self) if c is not None):
+            raise TQECException("Color must be 'x' or 'z'.")
+
+    def match(self, other: Color3D) -> bool:
+        """Check whether the color matches the other color."""
+        return all(
+            c1 is None or c2 is None or c1 == c2
+            for c1, c2 in zip(astuple(self), astuple(other))
+        )
+
+    def pop_color_at_direction(self, direction: Direction3D) -> Color3D:
+        """Replace the color at the given direction with None."""
+        return self.push_color_at_direction(direction, None)
+
+    def push_color_at_direction(
+        self, direction: Direction3D, color: str | None
+    ) -> Color3D:
+        """Set the color at the given direction."""
+        if direction == Direction3D.X:
+            return Color3D(color, self.y, self.z)
+        elif direction == Direction3D.Y:
+            return Color3D(self.x, color, self.z)
+        else:
+            return Color3D(self.x, self.y, color)
```

Comment:
Might be replaced with 
```suggestion
        colors = list(astuple(self))
        colors[direction.axis_index] = color
        return Color3D(*colors)
```